### ISSUE TITLE
 [Error: Unexpected FileStorage response file: null] on android 14 

### ISSUE THREAD

**Issue by YASH6004** [RonRadtke/react-native-blob-util#363]
Hi,

**- Steps to replicate the issue** 
1. Grant permission to the app
2.  download the data with the given code [it will download the zip file from the URL using cookie-based authentication ]
3. now delete the created folder from android > download > ${filename}  [ delete the whole folder ]
4. now try to download again it will throw this error but this will create the folder ${filename}

- <img width="626" alt="Screenshot 2024-05-24 at 1 52 25 PM" src="https://github.com/RonRadtke/react-native-blob-util/assets/102504196/d3bf8734-efe7-4709-9000-d2ee95fe6bb0">
- 
* RN version 
    "react": "18.1.0",
    "react-native": "0.70.6",
  
* android version 14  
   Device Samsung s23 ultra   

    
*here is my code sinnpet  '''
export const exportClientData = async (type: 'all' | 'single', client_id?: string) => {
  let filepath = '';
  let url = '';

  if (type === 'all') {
      url = `https://...`;
  } else if (type === 'single') {
      url = `https:///.......`;
  }

  // Get the app's cache directory
  const { fs } = ReactNativeBlobUtil;
  const cacheDir = Platform.OS === 'ios' ? fs.dirs.DocumentDir : fs.dirs.LegacyDownloadDir;

  // Define the filename
  const filename = 'filename';

  // Construct the full filepath
  if (type === 'all') {
      filepath = `${cacheDir}/${filename}/ClientData.zip`;
  } else if (type === 'single') {
      filepath = `${cacheDir}/${filename}/${client_id}.zip`;
  }

  try {
      // // Ensure the directory exists
      // const dirPath = `${cacheDir}/${filename}`;
      // const isDir = await fs.isDir(dirPath); // Check if the directory exists

      // if (!isDir) {
      //     await fs.mkdir(dirPath); // Create the directory if it does not exist
      // }

      // Download the file and save it to the cache directory
      const configOptions:any = Platform.select({
          ios: {
              fileCache: true,
              path: filepath,
              appendExt: 'zip',
          },
          android: {
              fileCache: true,
              path: filepath,
              appendExt: 'zip',
          },
      });

      const response = await ReactNativeBlobUtil.config(configOptions).fetch('GET', url, {
          "Accept": "*/*",
          "Accept-Encoding": "*/*",
      });
      console.log('response==>>', JSON.stringify(response, undefined, 4));
      return response;
  } catch (error) {
      console.error('error in downloading==>', error);
      return error;
  }
};
'''

**Comment by YASH6004** [comment]
any update anyone?

**Comment by topheroes** [comment]
@YASH6004 Facing the same problem. Did you solve it?

**Comment by valery-lavrik** [comment]
I'm getting a similar problem. Is there a solution?
The problem is reproduced when the file download request lasts more than 10 minutes.

**Comment by smfunder** [comment]
In my case I found that in the path parameter of config I should not include the "file://" prefix because that would make the library fail to create the container folder crashing the interceptor that later will be seen as "null" because it cannot be casted due the unhandled error.

**Comment by Eclipses-Saros** [comment]
I'm also getting a same problem. but somehow I've figured out one way to avoid this error: **DO NOT USE path option**.

```js
android: {
    fileCache: true, // <- if you enable this,
    path: filepath, // <- this cause a weird problem
    appendExt: 'zip',
},
```

once you have exception with `fetch` function, *somehow* BlobUtil fetch went completely wrong and cannot be fixed. I even checked with debugger, only know it will run the native code (`ReactNativeBlobUtilReq::run`) and suddenly pops out 'Error: Unexpected FileStorage response file: null'.

without path option, how can you move downloaded files to desire destination? use `ReactNativeBlobUtil.fs.cp` or `ReactNativeBlobUtil.MediaCollection.copyToMediaStore`. when `fileCache` option is set to true, you can get the cached file path.

```js
const res = await BlobUtil.config({
  fileCache: true
}).fetch();

await ReactNativeBlobUtil.MediaCollection.copyToMediaStore(
  {
    name: fileName,
    parentFolder: '',
    mimeType: type,
  },
  'Download',
  res.path(), // <- there it is
);

res.flush(); // <- if you complete file action, you can flush it
```

to summarize up:
1. `Error: Unexpected FileStorage response file: null` happens when you have `path` in config() and fetch() failed. then everything goes bad. 
2. try to copy cached file. this will prevent many unexpected errors.

**Comment by valery-lavrik** [comment]
> I'm also getting a same problem. but somehow I've figured out one way to avoid this error: **DO NOT USE path option**.
> 
> android: {
>     fileCache: true, // <- if you enable this,
>     path: filepath, // <- this cause a weird problem
>     appendExt: 'zip',
> },
> once you have exception with `fetch` _somehow_ BlobUtil fetch went completely wrong and cannot be fixed. I even checked with debugger, only know it will run the native code (`ReactNativeBlobUtilReq::run`) and suddenly pops out 'Error: Unexpected FileStorage response file: null'.
> 
> without path option, how can you move downloaded files to desire destination? use `ReactNativeBlobUtil.fs.cp` or `ReactNativeBlobUtil.MediaCollection.copyToMediaStore`. when `fileCache` option is set to true, you can get the cached file path.
> 
> const res = await BlobUtil.config({
>   fileCache: true
> }).fetch();
> 
> await ReactNativeBlobUtil.MediaCollection.copyToMediaStore(
>   {
>     name: fileName,
>     parentFolder: '',
>     mimeType: type,
>   },
>   'Download',
>   res.path(), // <- there it is
> );
> 
> res.flush(); // <- if you complete file action, you can flush it
> to summarize up:
> 
> 1. `Error: Unexpected FileStorage response file: null` happens when you have `path` in config() and fetch() failed. then everything goes bad.
> 2. try to copy cached file. this will prevent many unexpected errors.

Hopefully, after reading this answer, the developers will find and fix the problem.
In the end, it turned out to be easier for me to write my own module that just downloads a file...

**Comment by likhith-zep** [comment]
I faced this issue too, so updating my findings here if it helps

In our case, this error was because we were using [Chucker](https://github.com/ChuckerTeam/chucker) interceptor on our Http3 clients (not sure of the root cause, but disabling the interceptor appears to fix the issue though it's not something we wanted to go ahead with).
By using Android's DownloadManager instead of Http3, we were able to avoid this issue.

`addAndroidDownloads: {
        useDownloadManager: true, // since default http3 based download fails due to chucker
        path: downloadPath, // should not be documentDir (native error otherwise)
 }`

Add this to config, to use DownloadManager to carry out the download
